cranelift/codegen/src/isa/x64/inst/unwind.rs

```diff
@@ -1,36 +1,122 @@
-use super::Inst;
-use crate::isa::unwind::UnwindInfo;
-use crate::machinst::{UnwindInfoContext, UnwindInfoGenerator, UnwindInfoKind};
+use crate::isa::unwind::input::UnwindInfo;
+use crate::isa::x64::inst::{
+    args::{AluRmiROpcode, Amode, RegMemImm, SyntheticAmode},
+    regs, Inst,
+};
+use crate::machinst::{UnwindInfoContext, UnwindInfoGenerator};
 use crate::result::CodegenResult;
+use alloc::vec::Vec;
+use regalloc::Reg;
 
 #[cfg(feature = "unwind")]
-pub use self::systemv::create_cie;
-
-#[cfg(feature = "unwind")]
-mod systemv;
+pub(crate) mod systemv;
 
 pub struct X64UnwindInfo;
 
 impl UnwindInfoGenerator<Inst> for X64UnwindInfo {
-    #[allow(unused_variables)]
     fn create_unwind_info(
         context: UnwindInfoContext<Inst>,
-        kind: UnwindInfoKind,
-    ) -> CodegenResult<Option<UnwindInfo>> {
-        // Assumption: RBP is being used as the frame pointer for both calling conventions
-        // In the future, we should be omitting frame pointer as an optimization, so this will change
-        Ok(match kind {
-            #[cfg(feature = "unwind")]
-            UnwindInfoKind::SystemV => {
-                const WORD_SIZE: u8 = 8;
-                systemv::create_unwind_info(context, WORD_SIZE)?.map(UnwindInfo::SystemV)
-            }
-            #[cfg(feature = "unwind")]
-            UnwindInfoKind::Windows => {
-                //TODO winx64::create_unwind_info(context)?.map(|u| UnwindInfo::WindowsX64(u))
-                panic!();
+    ) -> CodegenResult<Option<UnwindInfo<Reg>>> {
+        use crate::isa::unwind::input::{self, UnwindCode};
+        let mut codes = Vec::new();
+        let word_size = 8u8;
+
+        for i in context.prologue.clone() {
+            let i = i as usize;
+            let inst = &context.insts[i];
+            let offset = context.insts_layout[i];
+
+            match inst {
```

Comment:
Hmm -- long-term, I think it might make more sense to generate the necessary `UnwindCode` sequence directly in prologue/epilogue generation, rather than reverse-engineering the VCode sequence; this feels somewhat fragile (we have to keep in mind this translation if we change the ABI code). Out of scope for this PR, but I wonder if you have any thoughts on that...

Reply:
"Reliable and Fast DWARF-Based Stack Unwinding" (https://hal.inria.fr/hal-02297690/document), actually states the how hard it is to keep the unwind info in sync while performing optimizations as one of it's reasons to generate unwind info from the actual assembly.

Reply:
Interesting; one thing to note is that the prologue generation in our case happens right before code emission, after any optimizations, but point taken that there is a potential for mismatch. I suppose as long as we are careful with any changes to the prologue code that the translation still works (though the tests look quite good here so that gives me confidence).